internal/vm/ioproxy.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package vm
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/firecracker-microvm/firecracker-containerd/internal"
+	"github.com/hashicorp/go-multierror"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+// IOProxy is an interface to a particular implementation for initializing
+// and copying the stdio of a process running in a VM. All its methods are
+// unexported as they are used only internally. The interface exists just
+// to give outside callers a choice of implementation when setting up a
+// process.
+type IOProxy interface {
+	// start should begin initialization of stdio proxying for the provided
+	// process. It returns two channels, one to indicate io initialization
+	// is completed and one to indicate io copying is completed.
+	start(proc *vmProc) (ioInitDone <-chan error, ioCopyDone <-chan error)
+}
+
+// IOConnector is function that begins initializing an IO connection (i.e.
+// vsock, FIFO, etc.). It returns two channels, one that should be published
+// to with the initialized ReadWriteCloser in the case of success and a
+// second with an error in the case of failure.
+//
+// The return of channels instead of plain values gives the implementation
+// the freedom to do synchronous setup, asynchronous setup or a mix of both.
+//
+// The provided context is canceled if the process fails to create/exec or
+// when the process exits after a successful create/exec.
+type IOConnector func(procCtx context.Context, logger *logrus.Entry) (<-chan io.ReadWriteCloser, <-chan error)
+
+// IOConnectorPair holds the read and write side of IOConnectors whose IO
+// should be proxied.
+type IOConnectorPair struct {
+	ReadConnector  IOConnector
+	WriteConnector IOConnector
+}
+
+func (connectorPair *IOConnectorPair) proxy(
+	proc *vmProc,
+	logger *logrus.Entry,
+	timeoutAfterExit time.Duration,
+) (ioInitDone <-chan error, ioCopyDone <-chan error) {
+	initDone := make(chan error)
+	copyDone := make(chan error)
+
+	inCh, inErrCh := connectorPair.ReadConnector(proc.ctx, logger.WithField("direction", "read"))
+	outCh, outErrCh := connectorPair.WriteConnector(proc.ctx, logger.WithField("direction", "write"))
```

Comment:
Just for my understanding. When error channel returns `nil`. Does it mean that `ReadWriteCloser` object is ok for use?

Reply:
> When error channel returns nil. Does it mean that ReadWriteCloser object is ok for use?

Yes, and if the error channel returns a non-nil object than the ReadWriteCloser should be nil. Just like the normal pattern of return values for go functions.

> In other words would this be an equivalent to chan Pair<ReadWriteCloser, error> ?

I don't think we'd be able to do that actually as the initialization finishing and the copying happening are two different events that happen at different times. You need two separate channels to communicate when those different events happen.